### ISSUE TITLE
Broken pagination due to incorrect example and deserializer bug

### ISSUE THREAD

**Issue by ppmathis** [microsoftgraph/msgraph-sdk-python#105]
## Issue Description
I attempted to use the latest release (`1.0.0a9`) of `msgraph-sdk` for listing all available groups of a MS365 tenant with pagination. A quick look into the documentation showed that there is an [example](https://github.com/microsoftgraph/msgraph-sdk-python/blob/main/docs/samples.md#paging-through-a-collection-of-items) available on how to paginate.

My first attempt, based on the documentation available, resulted in this code snippet:

```python
from msgraph.generated.groups.groups_request_builder import GroupsRequestBuilder
from msgraph.generated.models.group_collection_response import GroupCollectionResponse

async def test():
    query_params = GroupsRequestBuilder.GroupsRequestBuilderGetQueryParameters(
        select=['id', 'displayName', 'description'],
        top=2,
    )
    request_config = GroupsRequestBuilder.GroupsRequestBuilderGetRequestConfiguration(
        query_parameters=query_params,
    )

    groups = await client.groups.get(request_configuration=request_config)
    for group in groups.value:
        print(group.id, group.name, group.description)

    while groups.odata_next_link:
        request_info = client.groups.to_get_request_information(request_configuration=request_config)
        request_info.uri = groups.odata_next_link
        groups = request_adapter.send_async(request_info, GroupCollectionResponse)
        for group in groups.value:
            print(group.id, group.name, group.description)
```

Unfortunately this resulted in always only returning the first two groups instead of listing them all as expected. Further investigation on my side resulted in the following findings.

### Half-broken code for pagination example
The current code for the [pagination example](https://github.com/microsoftgraph/msgraph-sdk-python/blob/main/docs/samples.md#paging-through-a-collection-of-items) is somewhat broken:

- The line `print(msg.subject)` at the end of the snippet is not properly indented
- The call to `request_adapter.send_async()` does not use `await` despite calling an asynchronous function
- The call to `request_adapter.send_async()` fails due requiring error mappings. Passing an empty dictionary as a third argument works, but seems wrong given that [the usual `get()` method does specify error mappings](https://github.com/microsoftgraph/msgraph-sdk-python/blob/main/msgraph/generated/groups/groups_request_builder.py#L89-L92). Am I supposed to manually replicate them?
- The class `MessageCollectionResponse` is used without ever being imported anywhere and therefor does not work

I was able to come up with a solution for these issues myself, but this should really be fixed so that while pagination is still painful with the current library, it at least has a working example. I discovered though that even after all those fixes the `while` loop got never called due to `odata_next_link` being None, which leads me to the next issue.

### Broken field deserialization of `@odata.nextLink`
The class property `odata_next_link` is always `None`, which itself returns the class attribute `_odata_next_link` which is also `None`. Simulating the same kind of call via the Graph Explorer shows the expected `@odata.nextLink` key within the response however.

Dumping the contents of the first `groups` response using `__dict__` shows the following contents:

```
{'_additional_data': {'@odata.context': 'https://graph.microsoft.com/v1.0/$metadata#groups(id,displayName,description)', '@odata.next_link': 'https://graph.microsoft.com/v1.0/groups?$top=2&$select=id%2cdisplayName%2cdescription&$skiptoken=<redacted>'}, '_odata_count': None, '_odata_next_link': None, '_value': [<msgraph.generated.models.group.Group object at 0x10e4bde90>, <msgraph.generated.models.group.Group object at 0x10e4bde50>]}
```

This clearly shows that the response does contain the link to the next page, however it does not seem to be deserialized into `_odata_next_link`. I started debugging the library and noticed that the deserialization of `@odata.nextLink` is being configured as part of [get_field_deserializes](https://github.com/microsoftgraph/msgraph-sdk-python/blob/26313637d03c340a2f1648d31162f02a46c05adc/msgraph/generated/models/base_collection_pagination_count_response.py#L48-L57), which at a first glance seems to be correct.

I then however stumbled across `_assign_field_values` of the `kiota-serialization-json-python` library, which happens to [transform all keys into snake-case](https://github.com/microsoft/kiota-serialization-json-python/blob/90b630e03e469e83c1ed149b8ba3629b7edd4c2d/kiota_serialization_json/json_parse_node.py#L280-L281) before attempting to lookup the field deserializer. This means that while a deserializer is configured for `@odata.nextLink`, it is never considered after this transformation as Kiota is looking for `@odata.next_link`.

I confirmed this bug by manually editing `base_collection_pagination_count_response.py` and changing `@odata.nextLink` to `@odata.next_link`, after which the field was properly populated. Now I had a link to the next page, but still ended up with an infinite loop only returning the first two items over and over.

### Incorrect class attribute `uri` in example
After digging some more through the code, I discovered another flaw in the pagination example code. Within the while loop, it advises the user to override `request_info.uri` with the link to the next page to properly paginate through the results.

This is incorrect though, as the actual attribute is named `url` and overriding `uri` does nothing, therefor resulting in an endless loop. To get this working, the line has to be changed to `request_info.url = ...`.

## Working Example
If anyone else happens to struggle with pagination, this is an actual working example:

```python
from msgraph.generated.groups.groups_request_builder import GroupsRequestBuilder
from msgraph.generated.models.group_collection_response import GroupCollectionResponse

async def test():
    query_params = GroupsRequestBuilder.GroupsRequestBuilderGetQueryParameters(
        select=['id', 'displayName', 'description'],
        top=2,
    )
    request_config = GroupsRequestBuilder.GroupsRequestBuilderGetRequestConfiguration(
        query_parameters=query_params,
    )

    groups = await client.groups.get(request_configuration=request_config)
    for group in groups.value:
        print(group.id, group.name, group.description)

    while groups.odata_next_link:
        request_info = client.groups.to_get_request_information(request_configuration=request_config)
        request_info.url = groups.odata_next_link
        # Hint: Passing an empty error mapper might not be the ideal solution
        groups = await request_adapter.send_async(request_info, GroupCollectionResponse, {})
        for group in groups.value:
            print(group.id, group.name, group.description)
```

Please note that this still requires the bug regarding the `@odata.nextLink` deserialization to be fixed, as otherwise only the first page is being returned and the while loop will not be iterated even once.

## Feedback
I understand that this library is still an alpha version, but the current way of handling pagination is unbelievably painful. The addition of an optional keyword argument to calls like `.get()`, e.g. `next_link`, would already greatly simplify paginating through objects, even without any custom page iterators. That way the user would just have to call `.get(request_configuration=request_config, next_link=groups.odata_next_link)` to obtain the next page.

With the current implementation though, this is not possible and the user has to manually cobble various internal bits and pieces together for somehow obtaining a working request. This is a very tedious ordeal and combined with broken documentation probably sheer impossible for less experienced users.

**Comment by isvargasmsft** [comment]
Hello @ppmathis, thank you very much for raising this issue. 

We'll be building a page iterator for subsequent versions; it is in the roadmap, but it was not implemented in this alpha version. It will probably be a minor bump after the first GA. 

Please stay tuned for updates. 

That being said, thank you very much for your feedback about the current state. We'll look into this. 

CC: @shemogumbe, @samwelkanda

**Comment by akinnane** [comment]
Any updates on the pagination feature? It would be very useful to have instead of going through the pain @ppmathis described.

**Comment by scubbo** [comment]
Can confirm that ppmathis' example worked for me, thanks a ton! (If you are confused - as I was - about how to instantiate a `RequestAdapter`, note that it's a property on the `GraphServiceClient`; you don't need to create one afresh)

**Comment by teemuij** [comment]
I like the library and it has a ton of potential but pagination is pretty basic feature for an API library and is needed for achieving clean code. 

Catching the `odata_next_link` is already supported so my suggestion would be to add the functionality to the `xxxRequestBuilderGetRequestConfiguration` or `xxxRequestBuilderGetQueryParameters` classes.

**Comment by rpgeddam** [comment]
The python example documented on [this page](https://learn.microsoft.com/en-us/graph/paging?tabs=python) is very broken as well

```
# THE PYTHON SDK IS IN PREVIEW. FOR NON-PRODUCTION USE ONLY

graph_client = GraphServiceClient(credentials, scopes)

query_params = UsersRequestBuilder.UsersRequestBuilderGetQueryParameters(
		top = 5,
		skiptoken = "skiptoken=RFNwdAIAAQAAAD8...AAAAAAAA",
)

request_configuration = UsersRequestBuilder.UsersRequestBuilderGetRequestConfiguration(
query_parameters = query_params,
)

result = await graph_client.users.get(request_configuration = request_configuration)
```

The literal `skiptoken` appears twice in this line
```
skiptoken = "skiptoken=RFNwdAIAAQAAAD8...AAAAAAAA",
```
and as far as I can tell, `UsersRequestBuilderGetQueryParameters` [does not accept `skiptoken` as a field anyway](https://github.com/microsoftgraph/msgraph-sdk-python/blob/2fc6c0c68de98445c8d0e4f09bf9e441e4daf875/msgraph/generated/users/users_request_builder.py#L218-L237).

results in 
```
TypeError: UsersRequestBuilder.UsersRequestBuilderGetQueryParameters.__init__() got an unexpected keyword argument 'skiptoken'
```

**Comment by rpgeddam** [comment]
@ppmathis for your workaround, how do I make a `request_adapter` ?

**Comment by shemogumbe** [comment]
The pagination feature is done and [merged](https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/479), will be in the next release

**Comment by pl4nty** [comment]
@shemogumbe is this available in today's v1.3.0 release? I see that https://github.com/microsoftgraph/msgraph-sdk-python-core hasn't cut a release since the PR was merged
